src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -120,34 +119,115 @@ protected sealed override void InitializeWorker(AnalysisContext context)
+            var hasAttributeInAnalysisSpan = await ProcessSuppressMessageAttributesAsync(root, semanticModel, span,
+                idToSuppressMessageAttributesMap, suppressMessageAttributesToIsUsedMap, userExclusions, cancellationToken).ConfigureAwait(false);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Bail out if we have no pragma directives or SuppressMessageAttributes to analyze.
+            if (!hasPragmaInAnalysisSpan && !hasAttributeInAnalysisSpan)
+            {
+                return;
+            }
+
+            using var _8 = PooledHashSet<string>.GetInstance(out var idsToAnalyzeBuilder);
+            idsToAnalyzeBuilder.AddAll(idToPragmasMap.Keys);
+            idsToAnalyzeBuilder.AddAll(idToSuppressMessageAttributesMap.Keys);
+            var idsToAnalyze = idsToAnalyzeBuilder.ToImmutableHashSet();
+
+            // Compute all the reported compiler and analyzer diagnostics for diagnostic IDs corresponding to pragmas in the tree.
+            var (diagnostics, unhandledIds) = await GetReportedDiagnosticsForIdsAsync(
+                idsToAnalyze, root, semanticModel, compilationWithAnalyzers,
+                getSupportedDiagnostics, getIsCompilationEndAnalyzer, compilerDiagnosticIds, cancellationToken).ConfigureAwait(false);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Iterate through reported diagnostics which are suppressed in source through pragmas and mark the corresponding pragmas as used.
+            await ProcessReportedDiagnosticsAsync(diagnostics, tree, compilationWithAnalyzers, idToPragmasMap,
+                pragmasToIsUsedMap, idToSuppressMessageAttributesMap, suppressMessageAttributesToIsUsedMap, cancellationToken).ConfigureAwait(false);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Remove entries for unhandled diagnostic ids.
+            foreach (var id in unhandledIds)
+            {
+                foreach (var (pragma, _) in idToPragmasMap[id])
+                {
+                    pragmasToIsUsedMap.Remove(pragma);
+                }
+
+                if (idToSuppressMessageAttributesMap.TryGetValue(id, out var attributeNodes))
+                {
+                    foreach (var attributeNode in attributeNodes)
+                    {
+                        suppressMessageAttributesToIsUsedMap.Remove(attributeNode);
+                    }
+
+                    idToSuppressMessageAttributesMap.Remove(id);
+                }
+            }
+
+            // Finally, report the unnecessary suppressions.
+            var effectiveSeverity = severity.ToDiagnosticSeverity() ?? s_removeUnnecessarySuppressionDescriptor.DefaultSeverity;
+            ReportUnnecessarySuppressions(pragmasToIsUsedMap, sortedPragmasWithIds,
+                suppressMessageAttributesToIsUsedMap, reportDiagnostic, effectiveSeverity, compilationWithAnalyzers.Compilation);
```

Comment:
All of this is just existing code, which now shows up as new code as bunch of existing code was refactored into `ProcessPragmaDirectives`.